scripts/fetch_litecore.py

```diff
@@ -0,0 +1,309 @@
+    
+    Parameters
+    ----------
+    download_folder : str
+        The URL of the folder containing the variant to be downloaded
+    variant : str
+        The name of the variant (e.g. windows-win64)
+    debug: bool
+        If true, download the debug variant instead of the release one
+    output_base : str
+        The path to the base output directory for the downloaded files
+
+    Returns
+    -------
+    int
+        0 on success, 1 on failure (to be used for running tally of failures)
+    """
+
+    filename = filename_for_platform(variant, debug)
+    failed_count = 0
+    download_url = f"{download_folder}/{filename}"
+    download_path = calculate_download_path(variant, output_base)
+    print(f"--- Checking {filename}".ljust(70, '.'), end='')
+    try:
+        with urllib.request.urlopen(download_url):
+            print("200")
+    except HTTPError as e:
+        print(e.code)
+        failed_count = 1
+
+
+    print(f"\tDownload path: {download_path}")
+
+    return failed_count
+
+def main(variants, debug: bool, dry: bool, sha: str, ce: str, ee: str, output_path: str) -> int:
+    if not sha:
+        sha = calculate_sha(ce, ee)
+
+    download_folder = f"http://latestbuilds.service.couchbase.com/builds/latestbuilds/couchbase-lite-core/sha/{sha[0:2]}/{sha}"
+    print(f"--- Using URL {download_folder}/<filename>")
+    
+    failed_count = 0
+    for v in variants:
+        if dry:
+            failed_count += check_variant(download_folder, v, debug, output_path)
+        else:
+            failed_count += download_variant(download_folder, v, debug, output_path)
+
+    return failed_count
+
+def resolve_platform_path(path: str) -> Path:
+    """Calculates the absolute path to the folder containins platform extensions file
+    
+    Parameters
+    ----------
+    path : str
+        The path given as input to the script
+
+    Returns
+    -------
+    Path
+        The resolved, absolute path to the folder containing the platform extensions file
+    """
+
+    ret_val = Path(path)
+    if not ret_val.is_absolute():
+        ret_val = Path(os.path.dirname(__file__)).joinpath(ret_val).resolve()
+
+    if not ret_val.exists():
+        print(f"!!! {ret_val} does not exist, aborting...")
+        exit(-1)
+
+    return ret_val
+
+def import_platform_extensions(path: Path):
+    """Attempts to import the platform extensions
+    
+    Parameters
+    ----------
+    path : Path
+        The path to the folder containing the platform extensions file
+    """
+
+    try:
+        sys.path.insert(0, str(path.absolute()))
+        global subdirectory_for_variant
+        from platform_fetch import subdirectory_for_variant
+        global has_platform
+        has_platform = True
+    except ImportError:
+        print("!!! Unable to import platform extensions, falling back to <os>/<abi> for subdirectory...")
+
+def variant_to_pair(variant: str) -> Sequence[str]:
+    """Given a variant, splits it into an OS and ABI pair
```

Comment:
It took me a while to convince myself that this is, actually, ok.  I think the comment is wrong, though.  It is not converting to an OS and ABI pair... it is normalizing, very specifically, to LatestBuilds path names.

Once I realized that, though, it makes perfect sense.

I wonder if, instead of expressing this as code, it would make sense as a set of RE mappings.  That might, just possibly, be easier to understand, maintain and extend.

Just a thought.

Reply:
The RE would be pretty messy since the Windows ABI names are inconsistent and some platforms have no ABI at all (e.g. iOS is a fat binary with multiple architectures).  I wanted to standardize them to the official names since it will always be "wrong" for someone (i.e. Microsoft uses x64, Android uses x86_64).  For the OS name I wasn't sure what to do so I kept the latestbuilds name intact.  If something else would be better than I'd like to hear it.